File: go/tests/motor/drv8830_test/main.go
//go:build linux && !tinygo

// DRV8830 hardware test — Linux host.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/motor"
	"github.com/tuhde/Periph/go/periph/connection"
)

func envOr(k, def string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return def
}

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x60"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	m, err := motor.NewDRV8830Minimal(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "init:", err)
		os.Exit(2)
	}
	check("construct_minimal", m != nil)
	check("stop_no_error", m.Stop() == nil)

	full, err := motor.NewDRV8830Full(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "init:", err)
		os.Exit(2)
	}
	check("drive_forward_no_error", full.Drive(2.0) == nil)
	v, dir, err := full.ReadOutput()
	check("drive_forward_direction", err == nil && dir == motor.DRV8830Forward)
	check("drive_forward_voltage", v > 1.9 && v < 2.1)

	check("drive_reverse_no_error", full.Drive(-1.0) == nil)
	_, dir, err = full.ReadOutput()
	check("drive_reverse_direction", err == nil && dir == motor.DRV8830Reverse)

	check("brake_no_error", full.Brake() == nil)
	_, dir, err = full.ReadOutput()
	check("brake_direction", err == nil && dir == motor.DRV8830Brake)

	check("stop_no_error_full", full.Stop() == nil)
	_, dir, err = full.ReadOutput()
	check("stop_direction", err == nil && dir == motor.DRV8830Coast)

	check("clear_fault_no_error", full.ClearFault() == nil)
	f, err := full.ReadFault()
	check("clear_fault", err == nil && !f.Fault)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed > 0 {
		os.Exit(1)
	}
}

File: rust/tests/motor/drv8830_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::motor::{Drv8830Direction, Drv8830Full, Drv8830Minimal};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR").ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x60);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut motor = Drv8830Minimal::new(dev, addr).expect("init DRV8830");
    motor.stop().expect("stop");
    let dev = motor.release();

    let mut full = Drv8830Full::new(dev, addr).expect("init DRV8830 Full");
    full.drive(2.0).expect("drive");
    let out = full.read_output().expect("read_output");
    check_true!(out.direction == Drv8830Direction::Forward, "drive_forward_direction", passed, failed);
    check_true!((out.voltage - 2.0).abs() < 0.1, "drive_forward_voltage", passed, failed);

    full.drive(-1.0).expect("drive");
    check_true!(full.read_output().expect("read_output").direction == Drv8830Direction::Reverse, "drive_reverse_direction", passed, failed);

    full.brake().expect("brake");
    check_true!(full.read_output().expect("read_output").direction == Drv8830Direction::Brake, "brake_direction", passed, failed);

    full.stop().expect("stop");
    check_true!(full.read_output().expect("read_output").direction == Drv8830Direction::Coast, "stop_direction", passed, failed);

    full.clear_fault().expect("clear_fault");
    check_true!(!full.read_fault().expect("read_fault").fault, "clear_fault", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    if failed > 0 { std::process::exit(1); }
}
